Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp

```diff
@@ -1343,6 +1343,20 @@ void GraphicsLayerCA::setContentsDisplayDelegate(RefPtr<GraphicsLayerContentsDis
     noteLayerPropertyChanged(ContentsPlatformLayerChanged);
 }
 
+PlatformLayerIdentifier GraphicsLayerCA::setContentsToAsyncDisplayDelegate(ContentsLayerPurpose purpose)
+{
+    if (!m_contentsLayer) {
+        m_contentsLayer = createPlatformCALayer(PlatformCALayer::LayerTypeContentsProvidedLayer, this);
+        m_contentsLayerPurpose = purpose;
+        m_contentsLayer->setBackingStoreAttached(false);
```

Comment:
It's a bit weird to call this here. `setBackingStoreAttached(false)` is normally about things outside the viewport.

Reply:
This is necessary (or some other equivalent change), since we're pushing the buffer asynchronously to the UI-side layers.
The default value is true, and with that we'll end up trying to allocate a buffer for the RemoteLayerBackingStore and painting it, which fails (and we don't want it to work).

This works for normal delegated contents layers (like DOM <canvas> + WebGL), since that has special code paths in RemoteLayerBackingStore which bypasses the buffer allocation, and calls the delegate instead of paint().

Reply:
normally `drawsContent` is what determines if we allocate backing store for a layer, and it’s an error to call `setNeedsDisplay` on a layer which has `drawsContent = false`